### PR TITLE
Merge latest bits from `main` branch into extensions branch

### PR DESCRIPTION
**Pull request by jcouv**
There were some trivial conflicts (ErrorCode ans such). The non-trivial conflicts are in a separate commit.

### CODE REVIEW

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs [hunk 1]
```diff
@@ -16468,7 +16468,23 @@ public void M(int i)
...
+
+        verifier.VerifyDiagnostics();
+
+        verifier.VerifyIL("<top-level-statements-entry-point>", """
+{
+  // Code size       24 (0x18)
+  .maxstack  2
+  IL_0000:  newobj     "C..ctor()"
+  IL_0005:  ldftn      "void E.M(int)"
+  IL_000b:  newobj     "D..ctor(object, nint)"
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:16482]: This IL does look suspicious. We load an instance of `C` as a receiver, and method from `E`, which has no relationship with `C` from runtime point of view. Therefore, the IL doesn't look right. An addition of a prototype comment would be enough to unblock the merge.

File: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs [hunk 2]
```diff
@@ -17940,7 +17956,27 @@ public class C { }
...
+        verifier.VerifyIL("D.Main", """
+{
+  // Code size       29 (0x1d)
+  .maxstack  3
+  .locals init (int V_0) //i
+  IL_0000:  nop
+  IL_0001:  newobj     "D..ctor()"
+  IL_0006:  newobj     "C..ctor()"
+  IL_000b:  ldftn      "int E.M()"
+  IL_0011:  newobj     "System.Func<int>..ctor(object, nint)"
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:17974]: Similar issue.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 958)

File: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:16483]: Yes, this is expected for all instance scenarios at the moment. Added PROTOTYPE

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 959)

**Pull request merged**